### PR TITLE
Fix highlight with geometry generators

### LINKED ISSUE #48439: Highlight of identified feature not displaying correctly on "geometry generator" marker from different CRS
### What is the bug or the crash?

Highlight of identified feature (symbol outline in red by default) is not displaying correctly when symbol is created with geometry generator option.

Start point is correct (in project CRS), but endpoint seems to be used from layer CRS rather than in project CRS so we get weird red highlight visualization of identified feature.

See attached GPKG which includes project and data, demo in video below.

[QGIS Identify Highlight bug](http://www.youtube.com/watch?v=baJPCEe6NRw "QGIS Identify Highlight bug")

### Steps to reproduce the issue

See attached GPKG which includes project and data.
"Project -> Open From -> GeoPackage..."
[highlight_bug.zip](https://github.com/qgis/QGIS/files/8610408/highlight_bug.zip)

### Versions

QGIS version
3.24.2-Tisler
QGIS code revision
13c1a028
Qt version
5.15.2
Python version
3.9.5
GDAL/OGR version
3.4.2
PROJ version
9.0.0
EPSG Registry database version
v10.054 (2022-02-13)
GEOS version
3.10.2-CAPI-1.16.0
SQLite version
3.38.1
PDAL version
2.3.0
PostgreSQL client version
unknown
SpatiaLite version
5.0.1
QWT version
6.1.3
QScintilla2 version
2.11.5
OS version
Windows 10 Version 2009

Active Python plugins
EquirectangularViewer
1.4
ImportPhotos
3.0.3
mmqgis
2021.5.16
plaingeometryeditor
3.0.0
QuickWKT
3.1
db_manager
0.1.20
processing
2.12.99

### Supported QGIS version

- [X] I'm running a supported QGIS version according to the roadmap.

### New profile

- [X] I tried with a new QGIS profile

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by elpaso**
Fixes #48439

For geometry generators we need to generate the geometry using the layer's CRS and transform the generated
geometry to the canvas CRS

### CODE REVIEW

File: src/gui/qgshighlight.cpp [hunk 1]
```diff
@@ -168,6 +169,23 @@ void QgsHighlight::setSymbol( QgsSymbol *symbol, const QgsRenderContext &context
...
+      QgsFeature feature { mFeature };
+      feature.setGeometry( mOriginalGeometry );
+      ctx.setFeature( feature );
+      QgsGeometry geom { QgsExpression( generator->geometryExpression() ).evaluate( &ctx ).value<QgsGeometry>() };
+      const QgsCoordinateTransform ct { mMapCanvas->mapSettings().layerTransform( mLayer ) };
+      if ( ct.isValid() )
+      {
+        geom.transform( ct );
+      }
+      generator->setGeometryExpression( QStringLiteral( "geom_from_wkt('%1')" ).arg( geom.asWkt() ) );
```
Comment (m-kuhn) [inline src/gui/qgshighlight.cpp:186]: Is this line required?

Comment (m-kuhn) [inline src/gui/qgshighlight.cpp:186]: This looks very inefficient.
Can we somehow preserve the geometry as such?

**Comment by nyalldawson** [comment]
Isn't this a matter of making sure the correct transform is set in the render context? The rest should be handled automatically by the existing symbol layer logic.

**Comment by elpaso** [comment]
> 

This is what I tried first but no, the problem is that when the geometry generator symbol is generated it assumes that the geometry is in the layer's CRS while QgsHighlight transforms it early, I couldn't find another way: I also tried to set the geometry to the generated one but it changes the type from point to linestring and it doesn't render.

Also, unsetting the geometry in the feature in the QgsHighlight object ans setting the geometry does not work because you loose all the other style layers.

If this solution is not acceptable I'm happy to close this PR and leave the solution to other developers.

File: src/gui/qgshighlight.cpp (hunk 1 above)
Comment (elpaso) [reply src/gui/qgshighlight.cpp:186]: Tried without success.

**Comment by nyalldawson** [comment]
Let's just wait -- we can leave it open as a reminder to investigate alternative approaches and then merge before release

**Comment by elpaso** [comment]
> Let's just wait -- we can leave it open as a reminder to investigate alternative approaches and then merge before release

Sounds good to me.

**Review by strk — APPROVED** [review]
I'd say better slow than bogus. Rather, it would help to have an automated testcase, so it guards after correctness when it's time to refactor.

**Comment by strk** [comment]
I commented this PR, which did NOT automatically remove the Stale label. I've also *approved* this PR, which did not prevent the stale bot from closing it. @nyalldawson asked to keep it open, but the StaleBot still took over. They are coming for us ! ...

**Pull request closed without merging**